### PR TITLE
bring in class LogListHelper into MavlinkLogHandler

### PR DESCRIPTION
**Pull request by BazookaJoe1900**
this is followup to 'return with error from _log_request_list() in case of failed allocation #14451'

**Describe problem solved by this pull request**
The class _LogListHelper_ created and destroyed dynamically during logs download process. which mean some `new` and `delete` built in in _MavlinkLogHandler_ code. the existence of the class then checked, to decide the state of the downloader.
This PR remove the dynamic allocation. to ensure enough memory on creation of MavlinkLogHandler (meaning, creation of mavlink instance)

**Describe your solution**
The class LogListHelper removed, all its functions and variants moved into _MavlinkLogHandler_ and created when _MavlinkLogHandler_ allocated.
LOG_HANDLER_STATE::INACTIVE state is the state to indicate that the behavior as if _LogListHelper_ not exists

**Describe possible alternatives**
question:
now, on every creation of mavlink a MavlinkLogHandler allocated, with an about extra 150 bytes.
most of this memory is _current_log_filename[128].
https://github.com/PX4/Firmware/commit/661d538016b8a534a4d2ef0a900f487afce705fa#diff-e56b8d1bc66526d6f124b0bf662f02fdR105
is that acceptable?
it can be done that only _current_log_filename[] will be malloc dynamically, only if process of download is initiated.

**Test data / coverage**
tested proper behavior on FMUv5, download speed hasn't been changed and still ~330kBps

**Additional context**
more improvements can be done to the module. the commits that done keep the exact same behavior. and according to the response about this change, I can do more small/big fix on the process of downloading logs.

### CODE REVIEW

**Comment by BazookaJoe1900** [comment]
@bkueng, can you review?

File: src/modules/mavlink/mavlink_log_handler.cpp [hunk 1]
```diff
@@ -315,57 +305,43 @@ MavlinkLogHandler::_log_send_data()
...
-	, current_log_data_offset(0)
-	, current_log_data_remaining(0)
-	, current_log_filep(nullptr)
-{
-	_init();
-}
-
-//-------------------------------------------------------------------
-LogListHelper::~LogListHelper()
+void MavlinkLogHandler::_close_and_ulink_files(void)
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_log_handler.cpp:330]:
```suggestion
void MavlinkLogHandler::_close_and_unlink_files(void)
```

File: src/modules/mavlink/mavlink_log_handler.cpp [hunk 2]
```diff
@@ -315,57 +305,43 @@ MavlinkLogHandler::_log_send_data()
...
 {
-	if (current_log_filep) {
-		::fclose(current_log_filep);
+	if (_current_log_filep) {
+		::fclose(_current_log_filep);
 	}
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_log_handler.cpp:334]: Make sure to reset the state in here, like `_current_log_filep=nullptr`, `_next_entry=0`, ... (instead of doing it when starting a new request).

File: src/modules/mavlink/mavlink_log_handler.cpp [hunk 3]
```diff
@@ -315,57 +305,43 @@ MavlinkLogHandler::_log_send_data()
...
 
 	// Remove log data files (if any)
+	//TODO: is there need to check that kLogData and kTmpData exist?
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_log_handler.cpp:337]: Not necessarily, that call will just fail and set `errno`.

File: src/modules/mavlink/mavlink_log_handler.cpp [hunk 4]
```diff
@@ -109,24 +113,6 @@ MavlinkLogHandler::handle_message(const mavlink_message_t *msg)
 	}
 }
 
-//-------------------------------------------------------------------
-unsigned
-MavlinkLogHandler::get_size()
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_log_handler.cpp:114]: Why is this removed?

File: src/modules/mavlink/mavlink_log_handler.h [hunk 5]
```diff
@@ -105,15 +65,42 @@ class MavlinkLogHandler
 	unsigned get_size();
 
 private:
+	enum class LOG_HANDLER_STATE {
+		INACTIVE,
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_log_handler.h:69]: Can you clarify the difference between idle and inactive? Maybe use more descriptive names?

File: src/modules/mavlink/mavlink_log_handler.cpp (hunk 4 above)
Comment (BazookaJoe1900) [reply src/modules/mavlink/mavlink_log_handler.cpp:114]: @bkueng, because its simpler to use the data in ::send directly, get_size() function is not really needed

File: src/modules/mavlink/mavlink_log_handler.h (hunk 5 above)
Comment (BazookaJoe1900) [reply src/modules/mavlink/mavlink_log_handler.h:69]: @bkueng  I tried not to change the logic too much, not for now. so INACTIVE is similar to the case that the LogListHelper has not been created.
I added notes, and I know its a bit confusing, but without wider changes I don't want to change it much more.

**Comment by BazookaJoe1900** [comment]
@bkueng thanks for the review, I have made changes. and rebased.
can you re-review?

File: src/modules/mavlink/mavlink_log_handler.cpp [hunk 6]
```diff
@@ -315,47 +298,33 @@ MavlinkLogHandler::_log_send_data()
...
-	, current_log_data_offset(0)
-	, current_log_data_remaining(0)
-	, current_log_filep(nullptr)
-{
-	_init();
-}
-
-//-------------------------------------------------------------------
-LogListHelper::~LogListHelper()
+void MavlinkLogHandler::_close_and_unlink_files(void)
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_log_handler.cpp:323]: clang tidy complains about this:
```
/tmp/jenkins/workspace/PX4_Firmware_PR-14452/src/modules/mavlink/mavlink_log_handler.cpp:323:49: error: redundant void argument list in function definition [modernize-redundant-void-arg,-warnings-as-errors]

[2020-05-13T17:02:25.307Z] void MavlinkLogHandler::_close_and_unlink_files(void)
```

```suggestion
void MavlinkLogHandler::_close_and_unlink_files()
```

File: src/modules/mavlink/mavlink_log_handler.cpp (hunk 4 above)
Comment (bkueng) [reply src/modules/mavlink/mavlink_log_handler.cpp:114]: I thought you changed the logic, but it's correct.

File: src/modules/mavlink/mavlink_log_handler.h [hunk 7]
```diff
@@ -105,15 +65,43 @@ class MavlinkLogHandler
 	unsigned get_size();
 
 private:
+	enum class LOG_HANDLER_STATE {
+		INACTIVE,     //There is no active action of log handler
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_log_handler.h:69]: Can you change the naming convention:
```suggestion
	enum class LogHandlerState {
		Inactive,     //There is no active action of log handler
```

File: src/modules/mavlink/mavlink_log_handler.cpp (hunk 6 above)
Comment (BazookaJoe1900) [reply src/modules/mavlink/mavlink_log_handler.cpp:323]: > I assume you tested this?

For the last changes, I made them without Hardware, so didn't check yet. wait for any merge. I will re-test it in few days

**Comment by BazookaJoe1900** [comment]
@bkueng Changes done according to your notes, thanks.
Its also been tested on FMUv5 by downloading logs by the USB
